Remove static strings and symbols

justinmichaud opened the pull request:
#### 39a1ff098a4ebfd7b753243ba2d609fc5ef763eb
<pre>
Remove static strings and symbols
<a href="https://bugs.webkit.org/show_bug.cgi?id=256744">https://bugs.webkit.org/show_bug.cgi?id=256744</a>
rdar://109287900

Reviewed by Yusuke Suzuki.

We would like to make StringImpl*, JSString* and JSRopeString* fit into
32 bits.

Why?
- This has the potential to save memory and improve cache locallity.
- Previously, when Yusuke made JSRopeString smaller by squishing the ropes
    together, it was a 1% Speedometer regression. This will let us undo
    that change while keeping the size of JSRopeString small.

We need to:
1) Forbid precice allocation
2) Ensure 16-byte alignment

This patch does both of these things, and should be sufficient to enable
CompactPtr&lt;StringImpl&gt; to work on iOS (although this patch doesn&apos;t do that).

On macOS, pointers are 48-bit instead of 36-bit, so we have more work to do:
1) Make sure that all strings are allocated out of the same malloc region
2) Allocate a chunk of va space for them, and plumb the base pointer around

This patch does the first thing.

Isn&apos;t removing static strings going to be a memory regression?

Well, hopefully not. Today we still increment the ref count of static strings,
so the memory is dirtied. After applying this patch, the character data will
still be static, but the StringImpl itself (which remember, was always dirty)
will be placed next to its bretheren.

This patch might introduce some concurrency bugs though. By inspection I
have tried to match the existing invariants, but I don&apos;t know if I have
spotted everything.

* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::defaultConstructorSourceCode):
* Source/JavaScriptCore/builtins/BuiltinExecutables.h:
* Source/JavaScriptCore/builtins/BuiltinNames.cpp:
(JSC::Symbols::initializeStaticSymbols):
(JSC::BuiltinNames::BuiltinNames):
(): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::JSString::destroy): Deleted.
(JSC::JSString::subspaceFor): Deleted.
(JSC::JSString::uninitializedValueInternal const): Deleted.
(JSC::JSString::valueInternal const): Deleted.
(JSC::JSString::JSString): Deleted.
(JSC::JSString::finishCreation): Deleted.
(JSC::JSString::create): Deleted.
(JSC::JSString::createHasOtherOwner): Deleted.
(JSC::JSString::toBoolean const): Deleted.
(JSC::JSString::canGetIndex): Deleted.
(JSC::JSString::offsetOfValue): Deleted.
(JSC::JSString::isRope const): Deleted.
(JSC::JSString::fiberConcurrently const): Deleted.
* Source/JavaScriptCore/runtime/SmallStrings.cpp:
* Source/JavaScriptCore/runtime/SmallStrings.h:
(JSC::SmallString::initializeJSStaticStrings):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::ensureTerminationException):
(): Deleted.
* Source/WTF/wtf/CompactPtr.h:
(WTF::CompactPtr::operator!=):
(WTF::CompactPtr::storage const):
(WTF::CompactPtr::storage):
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):
* Source/WTF/wtf/URL.cpp:
(WTF::aboutBlankURL):
(WTF::aboutSrcDocURL):
(): Deleted.
* Source/WTF/wtf/text/AtomString.cpp:
(WTF::nullAtom):
(WTF::emptyAtom):
(WTF::AtomString::initializeStrings):
(): Deleted.
* Source/WTF/wtf/text/AtomString.h:
(WTF::StaticAtomString::StaticAtomString): Deleted.
(WTF::nullAtom): Deleted.
(WTF::emptyAtom): Deleted.
* Source/WTF/wtf/text/AtomStringImpl.cpp:
* Source/WTF/wtf/text/AtomStringImpl.h:
(WTF::ValueCheck&lt;AtomStringImpl::checkConsistency): Deleted.
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::createStaticStringImplWithoutCopying):
(WTF::StringImpl::createStaticStringImpl):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::createStaticStringImplWithoutCopying):
(WTF::StringImpl::empty):
(WTF::StringImpl::initializeEmptyString):
(WTF::StringImpl::StringImpl):
(WTF::StringImpl::cost const):
(WTF::ValueCheck&lt;StringImpl::checkConsistency): Deleted.
(WTF::StringImpl::StaticStringImpl::StaticStringImpl): Deleted.
(WTF::StringImpl::StaticStringImpl::operator StringImpl&amp;): Deleted.
* Source/WTF/wtf/text/SymbolImpl.cpp:
(WTF::SymbolImpl::createStatic):
(WTF::PrivateSymbolImpl::createStatic):
* Source/WTF/wtf/text/SymbolImpl.h:
(WTF::SymbolImpl::create):
(WTF::SymbolImpl::createStatic):
(WTF::ValueCheck&lt;SymbolImpl::checkConsistency): Deleted.
* Source/WTF/wtf/text/UniquedStringImpl.h:
(WTF::ValueCheck&lt;UniquedStringImpl::checkConsistency): Deleted.
(): Deleted.
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::nullString):
(WTF::emptyString):
(WTF::String::initializeStrings):
* Source/WTF/wtf/text/WTFString.h:
(WTF::StaticString::StaticString): Deleted.
(WTF::nullString): Deleted.
(WTF::emptyString): Deleted.
* Source/WebCore/bindings/scripts/StaticString.pm:
(GenerateStrings):
(GenerateStringAsserts):
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::matchingFacesExcludingPreinstalledFonts):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::m_version):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::maskedURLForBindings):
(WebCore::Document::maskedURLStringForBindings): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::resolveURLStringIfNeeded const):
* Source/WebCore/dom/make_names.pl:
(printTagNameCppFile):
(printDefinitions):
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::realizeFallbackRangesAt):
* Source/WebCore/style/StyleResolveForFontRaw.cpp:
(WebCore::Style::resolveForFontRaw):
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceMain):
* Tools/TestWebKitAPI/Tests/WTF/StringImpl.cpp:
(TestWebKitAPI::TEST):
(): Deleted.

Canonical link: <a href="https://commits.webkit.org/264265@main">https://commits.webkit.org/264265@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/422256def58aaa8f4b32e5a83d62c428e6052911

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7125 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7370 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7549 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8739 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7349 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8632 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7302 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10247 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7252 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6564 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8845 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5262 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6482 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14230 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6035 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6926 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6581 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9455 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6707 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7061 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5775 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/7262 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6421 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1684 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1702 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10620 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7463 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6804 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1824 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

ddegazio commented on Source/JavaScriptCore/builtins/BuiltinNames.cpp:
```diff
@@ -37,30 +37,47 @@
+LazyNeverDestroyed<SymbolImpl*> intlLegacyConstructedSymbol;
+
+#define DEFINE_BUILTIN_PRIVATE_NAMES(name) LazyNeverDestroyed<SymbolImpl*> name##PrivateName;
+JSC_FOREACH_BUILTIN_FUNCTION_NAME(DEFINE_BUILTIN_PRIVATE_NAMES)
```
Nit: if the other `JSC_FOREACH_...` macro calls are indented, these should probably be indented too. Or maybe don't indent any of them - your call.

ddegazio commented on Source/JavaScriptCore/runtime/SmallStrings.h:
```diff
@@ -49,6 +50,30 @@ class StringImpl;
+#undef DECLARE_STATIC_STRING_IMPL
+
+#define INITIALIZE_STRING(name, characters) \
+    s_##name.construct(&StringImpl::createStaticStringImplWithoutCopying(characters, sizeof(characters) / sizeof(char) - 1 /* skip null byte */).leakRef());
```
These `sizeof()` uses make me a little worried...what are the chances we accidentally pass in a non-literal value at some point? Maybe we can add a 
```
static_assert(std::is_array_v<std::remove_reference_t<decltype(characters)>>)
``` 
to guarantee the argument hasn't decayed to a pointer.

justinmichaud replied:
Anywhere that calls createStaticStringImplWithoutCopying needs to either be inside a call_once, or needs a fence

ddegazio commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -642,8 +635,7 @@ template<> struct DefaultHash<PackedPtr<StringImpl>>;
 #define MAKE_STATIC_STRING_IMPL(characters) ([] { \
-        static StaticStringImpl impl(characters); \
-        return &impl; \
+        return &StringImpl::createStaticStringImplWithoutCopying(characters, sizeof(characters) / sizeof(char) - 1).get(); \
```
Ditto.

achristensen07 commented on Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
```diff
@@ -37,6 +37,7 @@ namespace WebKit {
 
 const Vector<RefPtr<API::Object>>& WebPreferences::features()
 {
+    WTF::initialize(); // We use WTF strings.
```
This looks like it may be in the wrong place.  I would expect InitializeWebKit2() to take care of this.

justinmichaud replied:
This gets called from:
```7   WebKit                        	       0x1083ac780 WebKit::WebPreferences::features() + 152 (WebPreferencesFeatures.cpp:42)
8   WebKit                        	       0x107ff6bf0 WebKit::WebPreferences::experimentalFeatures()::$_74::operator()() const + 36 (WebPreferences.cpp:85)
9   WebKit                        	       0x1084dc5a8 WebKit::WebPreferences::experimentalFeatures() + 64 (WebPreferences.cpp:83)
10  WebKit                        	       0x1084801dc +[WKPreferences(WKPrivate) _experimentalFeatures] + 24 (WKPreferences.mm:564)
11  MiniBrowser                   	       0x104a53f14 -[SettingsController _populateMenu:] + 1772 (SettingsController.m:207)
12  MiniBrowser                   	       0x104a537b4 -[SettingsController initWithMenu:] + 436 (SettingsController.m:109)
13  MiniBrowser                   	       0x104a50844 -[BrowserAppDelegate awakeFromNib] + 40 (AppDelegate.m:83)
14  CoreFoundation                	       0x18701a5f0 -[NSSet makeObjectsPerformSelector:] + 176
15  AppKit                        	       0x18a22631c -[NSIBObjectData nibInstantiateWithOwner:options:topLevelObjects:] + 824
16  AppKit                        	       0x18a21be9c loadNib + 340
17  AppKit                        	       0x18a21b4cc +[NSBundle(NSNibLoading) _loadNibFile:nameTable:options:withZone:ownerBundle:] + 560
18  AppKit                        	       0x18a21b1d0 -[NSBundle(NSNibLoading) loadNibNamed:owner:topLevelObjects:] + 180
19  AppKit                        	       0x18a21aff4 +[NSBundle(NSNibLoading) loadNibNamed:owner:] + 300
20  AppKit                        	       0x18a20e750 NSApplicationMain + 488
21  MiniBrowser                   	       0x104a535a8 main + 48 (main.m:33)
22  dyld                          	       0x186beff28 start + 2236```

justinmichaud replied:
Offline Alex suggested that we should be calling InitializeWebkit2 from anywhere that calls into _experimentalFeatures and its kin instead

justinmichaud replied:
done

achristensen07 commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
```diff
@@ -243,6 +243,7 @@ int XPCServiceMain(int, const char**)
             }
         }
 #endif
+        WTF::initialize(); // We use WTF strings.
```
This should probably be removed.  The WTF::String use in the following few lines should probably be changed to use NSString and c strings instead.  It should be pretty trivial.

justinmichaud replied:
Done, will upload shortly

rniwa commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
```diff
@@ -243,11 +243,13 @@ int XPCServiceMain(int, const char**)
+        if (!webKitBundleVersion || !expectedBundleVersion || strncmp(webKitBundleVersion, expectedBundleVersion, strlen(expectedBundleVersion)) != 0) {
+            const int msgLen = strlen("WebKit framework version mismatch: %s != %s\n") + strlen(webKitBundleVersion) + strlen(expectedBundleVersion);
+            char* msg = (char*) malloc(sizeof(char) * msgLen);
+            snprintf(msg, msgLen, "WebKit framework version mismatch: %s != %s\n", webKitBundleVersion, expectedBundleVersion);
```
Why are we changing this to snprintf?

justinmichaud replied:
WTF isn't initialized yet here, and we can't link against it yet.

rniwa replied:
Oh, I see.

rniwa reviewed (approved):
WebCore/WebKit changes look okay to me.

Constellation reviewed (approved):
r=me

justinmichaud committed:
Remove static strings and symbols
https://bugs.webkit.org/show_bug.cgi?id=256744
rdar://109287900

Reviewed by Yusuke Suzuki.

We would like to make StringImpl*, JSString* and JSRopeString* fit into
32 bits.

Why?
- This has the potential to save memory and improve cache locallity.
- Previously, when Yusuke made JSRopeString smaller by squishing the ropes
    together, it was a 1% Speedometer regression. This will let us undo
    that change while keeping the size of JSRopeString small.

We need to:
1) Forbid precice allocation
2) Ensure 16-byte alignment

This patch does both of these things, and should be sufficient to enable
CompactPtr<StringImpl> to work on iOS (although this patch doesn't do that).

On macOS, pointers are 48-bit instead of 36-bit, so we have more work to do:
1) Make sure that all strings are allocated out of the same malloc region
2) Allocate a chunk of va space for them, and plumb the base pointer around

This patch does the first thing.

Isn't removing static strings going to be a memory regression?

Well, hopefully not. Today we still increment the ref count of static strings,
so the memory is dirtied. After applying this patch, the character data will
still be static, but the StringImpl itself (which remember, was always dirty)
will be placed next to its bretheren.

This patch might introduce some concurrency bugs though. By inspection I
have tried to match the existing invariants, but I don't know if I have
spotted everything.

* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::defaultConstructorSourceCode):
* Source/JavaScriptCore/builtins/BuiltinExecutables.h:
* Source/JavaScriptCore/builtins/BuiltinNames.cpp:
(JSC::Symbols::initializeStaticSymbols):
(JSC::BuiltinNames::BuiltinNames):
(): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::JSString::destroy): Deleted.
(JSC::JSString::subspaceFor): Deleted.
(JSC::JSString::uninitializedValueInternal const): Deleted.
(JSC::JSString::valueInternal const): Deleted.
(JSC::JSString::JSString): Deleted.
(JSC::JSString::finishCreation): Deleted.
(JSC::JSString::create): Deleted.
(JSC::JSString::createHasOtherOwner): Deleted.
(JSC::JSString::toBoolean const): Deleted.
(JSC::JSString::canGetIndex): Deleted.
(JSC::JSString::offsetOfValue): Deleted.
(JSC::JSString::isRope const): Deleted.
(JSC::JSString::fiberConcurrently const): Deleted.
* Source/JavaScriptCore/runtime/SmallStrings.cpp:
* Source/JavaScriptCore/runtime/SmallStrings.h:
(JSC::SmallString::initializeJSStaticStrings):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::ensureTerminationException):
(): Deleted.
* Source/WTF/wtf/CompactPtr.h:
(WTF::CompactPtr::operator!=):
(WTF::CompactPtr::storage const):
(WTF::CompactPtr::storage):
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):
* Source/WTF/wtf/URL.cpp:
(WTF::aboutBlankURL):
(WTF::aboutSrcDocURL):
(): Deleted.
* Source/WTF/wtf/text/AtomString.cpp:
(WTF::nullAtom):
(WTF::emptyAtom):
(WTF::AtomString::initializeStrings):
(): Deleted.
* Source/WTF/wtf/text/AtomString.h:
(WTF::StaticAtomString::StaticAtomString): Deleted.
(WTF::nullAtom): Deleted.
(WTF::emptyAtom): Deleted.
* Source/WTF/wtf/text/AtomStringImpl.cpp:
* Source/WTF/wtf/text/AtomStringImpl.h:
(WTF::ValueCheck<AtomStringImpl::checkConsistency): Deleted.
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::createStaticStringImplWithoutCopying):
(WTF::StringImpl::createStaticStringImpl):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::createStaticStringImplWithoutCopying):
(WTF::StringImpl::empty):
(WTF::StringImpl::initializeEmptyString):
(WTF::StringImpl::StringImpl):
(WTF::StringImpl::cost const):
(WTF::ValueCheck<StringImpl::checkConsistency): Deleted.
(WTF::StringImpl::StaticStringImpl::StaticStringImpl): Deleted.
(WTF::StringImpl::StaticStringImpl::operator StringImpl&): Deleted.
* Source/WTF/wtf/text/SymbolImpl.cpp:
(WTF::SymbolImpl::createStatic):
(WTF::PrivateSymbolImpl::createStatic):
* Source/WTF/wtf/text/SymbolImpl.h:
(WTF::SymbolImpl::create):
(WTF::SymbolImpl::createStatic):
(WTF::ValueCheck<SymbolImpl::checkConsistency): Deleted.
* Source/WTF/wtf/text/UniquedStringImpl.h:
(WTF::ValueCheck<UniquedStringImpl::checkConsistency): Deleted.
(): Deleted.
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::nullString):
(WTF::emptyString):
(WTF::String::initializeStrings):
* Source/WTF/wtf/text/WTFString.h:
(WTF::StaticString::StaticString): Deleted.
(WTF::nullString): Deleted.
(WTF::emptyString): Deleted.
* Source/WebCore/bindings/scripts/StaticString.pm:
(GenerateStrings):
(GenerateStringAsserts):
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::matchingFacesExcludingPreinstalledFonts):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::m_version):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::maskedURLForBindings):
(WebCore::Document::maskedURLStringForBindings): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::resolveURLStringIfNeeded const):
* Source/WebCore/dom/make_names.pl:
(printTagNameCppFile):
(printDefinitions):
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::realizeFallbackRangesAt):
* Source/WebCore/style/StyleResolveForFontRaw.cpp:
(WebCore::Style::resolveForFontRaw):
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceMain):
* Tools/TestWebKitAPI/Tests/WTF/StringImpl.cpp:
(TestWebKitAPI::TEST):
(): Deleted.

Canonical link: https://commits.webkit.org/264265@main
